pallet-chainlink-feed/src/lib.rs

```diff
@@ -0,0 +1,1345 @@
+#[derive(Clone, Encode, Decode, Default, Eq, PartialEq, RuntimeDebug)]
+pub struct OracleMeta<AccountId: Parameter, Balance: Parameter> {
+	withdrawable: Balance,
+	admin: AccountId,
+	pending_admin: Option<AccountId>,
+}
+pub type OracleMetaOf<T> = OracleMeta<<T as frame_system::Trait>::AccountId, BalanceOf<T>>;
+
+/// Meta data tracking the oracle status for a feed.
+#[derive(Clone, Encode, Decode, Default, Eq, PartialEq, RuntimeDebug)]
+pub struct OracleStatus<RoundId: Parameter, Value: Parameter> {
+	starting_round: RoundId,
+	ending_round: Option<RoundId>,
+	last_reported_round: Option<RoundId>,
+	last_started_round: Option<RoundId>,
+	latest_submission: Option<Value>,
+}
+pub type OracleStatusOf<T> = OracleStatus<<T as Trait>::RoundId, <T as Trait>::Value>;
+
+impl<R, V> OracleStatus<R, V>
+where
+	R: Parameter + Default, // RoundId
+	V: Parameter + Default, // Value
+{
+	/// Create a new oracle status with the given `starting_round`.
+	fn new(starting_round: R) -> Self {
+		Self {
+			starting_round,
+			..Default::default()
+		}
+	}
+}
+
+/// Used to store round requester permissions for accounts.
+#[derive(Clone, Encode, Decode, Default, Eq, PartialEq, RuntimeDebug)]
+pub struct Requester<RoundId: Parameter> {
+	delay: RoundId,
+	last_started_round: Option<RoundId>,
+}
+pub type RequesterOf<T> = Requester<<T as Trait>::RoundId>;
+
+/// Round data as served by the `FeedInterface`.
+#[derive(Clone, Encode, Decode, Default, Eq, PartialEq, RuntimeDebug)]
+pub struct RoundData<BlockNumber: Parameter, RoundId: Parameter, Value: Parameter> {
+	pub started_at: BlockNumber,
+	pub answer: Value,
+	pub updated_at: BlockNumber,
+	pub answered_in_round: RoundId,
+}
+pub type RoundDataOf<T> =
+	RoundData<<T as frame_system::Trait>::BlockNumber, <T as Trait>::RoundId, <T as Trait>::Value>;
+
+/// Possible error when converting from `Round` to `RoundData`.
+#[derive(Clone, Encode, Decode, Eq, PartialEq, RuntimeDebug)]
+pub enum RoundConversionError {
+	MissingField,
+}
+
+impl<B, R, V> TryFrom<Round<B, R, V>>
+	for RoundData<B, R, V>
+where
+	B: Parameter, // BlockNumber
+	R: Parameter, // RoundId
+	V: Parameter, // Value
+{
+	type Error = RoundConversionError;
+
+	fn try_from(r: Round<B, R, V>) -> Result<Self, Self::Error> {
+		if r.answered_in_round.is_none() || r.answer.is_none() || r.updated_at.is_none() {
+			return Err(RoundConversionError::MissingField);
+		}
+		Ok(Self {
+			started_at: r.started_at,
+			answer: r.answer.unwrap(),
+			updated_at: r.updated_at.unwrap(),
+			answered_in_round: r.answered_in_round.unwrap(),
+		})
+	}
+}
+
+/// Trait for interacting with the feeds in the pallet.
+pub trait FeedOracle {
+	type FeedId: Parameter + BaseArithmetic;
+	type Feed: FeedInterface;
+
+	/// Return the interface for the given feed if it exists.
+	fn feed(id: Self::FeedId) -> Option<Self::Feed>;
+}
+
+/// Trait for interacting with a particular feed.
+pub trait FeedInterface {
```

Comment:
I wonder if there is any benefit to keeping all the types generic, rather than `trait FeedInterface<T: Config>`, given that you don't want `FeedInterface` to be vastly reusable and dynamic like substrate?

Reply:
Right, it's unlikely someone will use the interface with a different pallet.

Reply:
I tried this and AFAICT it sucks for integration. It's much better to be able to get the types returned from the traits functions as part of the trait.
Getting a `T` that implements `pallet_chainlink_feed::Trait` is work, too.